### ISSUE TITLE
Non-normalized versions of measures

### ISSUE THREAD

**Issue by juliohm** [alan-turing-institute/MLJ.jl#445]
**Is your feature request related to a problem? Please describe.**
Currently, measures such as `misclassification_rate` accept weights, but these weights are used to weight the examples as well as normalise the sum, i.e. `sum(w .* (yhat .!= y)) / sum(w)`. This is useful in some contexts, but not very useful for cost-sensitive learning where the weighted version is not normalised `sum(w .* (yhat .!= y))`. The latter version is more general since we can always manually force the weights to sum up to 1.

**Describe the solution you'd like**
By default the measures shouldn't divide the sum over examples by the sum of the weights. The user perhaps could provide a keyword argument like `normalize=true` if he/she desires to force the normalisation.

**Describe alternatives you've considered**
I am currently "unweighting" the result of measures because that is the expression most commonly used in cost-sensitive learning approaches and transfer learning. That is, I first call `loss(yhat, y, w)` and then multiply by `sum(w)`.

**Additional context**
This has been discussed over slack, and we all seem to agree the normalisation isn't needed.

**Comment by ablaom** [comment]
This sounds reasonable to me. @juliohm Are you happy to make a PR for the suggested changes?

**Comment by juliohm** [comment]
Will try to work on it in the following weeks @ablaom , thanks for the green light.

**Comment by juliohm** [comment]
I've started working on this. We really need a better (more efficient + more clean) API for loss functions a la LossFunctions.jl. I will submit a PR with a major refactoring of the code in MLJBase.jl, and we can continue the discussion there.

The idea is to introduce a single abstract `Loss` type, make all losses subtypes of this type, and combine the traits of LossFunctions.jl with the traits of MLJBase.jl. Ideally, we should be able to concentrate most of the code for losses (a general aggregation loop) in a single place, and only specialise the inner parts for the different loss types.

**Comment by ablaom** [comment]
Thanks for taking this on. While I am confident you can make some great improvements, I advise you to proceed with caution and would advise against a full-blown refactor without further discussion. **The suggestion you have made in this PR is straight-forward and orthogonal  to any other issues** you appear to have, as far as I can see. 

I don't agree the LossFunctions.jl is that clean. For a start, the API for classification is very specific to the binary case. Also, the probabilistic/deterministic distinction is not made explicit. All their regression losses are deterministic; all their classification losses are (sort of) probabilistic.  I say "sort of" because the probabilities are allowed to fall outside of [0, 1] range (after rescaling from their representation between -1 and 1).

I don't like your suggestion to separate measures into losses and scores at the type level. (Indeed, I don't see a need for measure types at all.) The `orientation` trait works well and is already heavily integrated into the the tuning API, etc,  (And there was  no oher support for your proposal to make this change at https://github.com/alan-turing-institute/MLJBase.jl/issues/51.)

 The LossFunctions does not accomodate loss functions that are also feature dependent (eg, think of a loss function for grocery items that weights the perishable goods higher than the others  - requires a custom loss depending on a feature value). 

I do not think measures should need to subtype a common abstract type. To what end?
While MLJ does have a ` is_measure` trait, even this is only needed to make the `info` method work for those measures and is not critical. For this reason it is very easy for users to define measures (please read [documentation](https://alan-turing-institute.github.io/MLJ.jl/dev/performance_measures/) on performance  measures).  

Whatever the changes you make, please keep in tact the `value(measure, yhat, X, y, w)` interface, as all the meta-algorithms depend on that. Again, see the docs for details. 

Thanks.

**Comment by ablaom** [comment]
Perhaps if you have aspects of LossFunctions that you would like duplicated in MLJ, you can open a new issue outlining them. Thanks.

**Comment by juliohm** [comment]
Thank you for the follow up. I will submit a first PR containing just the normalisation fix, and then we can consider a separate PR with the changes I have in mind. Neither LossFunctions.jl nor MLJBase.jl are perfect. There are lots of code repetition, and many type instabilities. Furthermore, some of the code may not work on GPUs, nor be automatically differentiable by Zygote, which is a bummer for people interested in loss functions for deep neural networks for example.

**Comment by juliohm** [comment]
Users can unweight the results manually if they want. Closing the issue.

**Comment by ablaom** [comment]
@juliohm FYI, normalisation was removed already in MLJBase 0.12.0. See release notes.
